### PR TITLE
Vectorize the softmax calculation when not along the last dim

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Currently, if we do softmax which are not along the last dim, the calculation will fall to a [scalar version](https://github.com/pytorch/pytorch/blob/d417a094f398f1c4efd7f818b14b8471a597fbcc/aten/src/ATen/native/SoftMax.cpp#L14-L64).  And we find actually we have the chance to vectorize the calculation along the inner_size dim.

Changes we made:

- Use vectorized softmax_kernel instead of host_softmax when not along the last dim.

Performance data on 28 cores' Intel 8280 CPU when the Input size is [32, 81, 15130] and do softmax along the second dim(81).

- FP32 Baseline: 24.67 ms
- FP32 optimized: 9.2 ms

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
Hi @yf225, I have met some `TestCppApiParity` UT fails in this PR as  
```
test_torch_nn_Softmax2d - TestCppApiParity
test_cpp_api_parity.py
```
However, I can pass the test in my local env with command: `python test/test_cpp_api_parity.py TestCppApiParity.test_torch_nn_Softmax2d`

Do you know how can I reproduce the fails in the UT?

**Comment by leslie-fang-intel** [comment]
> Hello @leslie-fang-intel, thank you for your PR! Could you also post some benchmarks for other (smaller and larger) shapes and dimensions including benchmarks on some different dtypes that are also being implemented here? Also, how does this scale with the number of CPU cores available? Thank you.

Hi @cpuhrsch, thanks for the suggestions. The [original implementation](https://github.com/pytorch/pytorch/blob/d417a094f398f1c4efd7f818b14b8471a597fbcc/aten/src/ATen/native/SoftMax.cpp#L143-L145) only support dtype with fp32 and fp64. So for this PR, I think we should keep within support of these 2 dtypes. And as for the extra support of BF16, I can open up a following PR if this one looks ok.

Here is the performance data I collect on 28cores' Intel 8380H CPU with different shapes and core numbers:




<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
<link id="Main-File" rel="Main-File" href="file:///C:/Users/jiahaofa/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel="File-List" href="file:///C:/Users/jiahaofa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>





Shape | Dtype | Dim | Cores | Base Performance(ms) | Optimized Performance(ms)
-- | -- | -- | -- | -- | --
[32, 81, 15130] | fp32 | -2 | 28 | 23.7605 | 8.6088
[32, 16, 15130] | fp32 | -2 | 28 | 3.8581 | 0.8681
[1, 81, 15130] | fp32 | -2 | 28 | 0.5321 | 0.0821
[1, 16, 15130] | fp32 | -2 | 28 | 0.1178 | 0.0221
[32, 81, 20] | fp32 | -2 | 28 | 0.0316 | 0.0261
[32, 16, 20] | fp32 | -2 | 28 | 0.0261 | 0.0265
[1, 81, 20] | fp32 | -2 | 28 | 0.0409 | 0.0366
[1, 16, 20] | fp32 | -2 | 28 | 0.0137 | 0.0129
[32, 81, 16] | fp32 | -2 | 28 | 0.0296 | 0.0274
[32, 16, 16] | fp32 | -2 | 28 | 0.015 | 0.0118
[1, 81, 16] | fp32 | -2 | 28 | 0.034 | 0.0319
[1, 16, 16] | fp32 | -2 | 28 | 0.0121 | 0.0117
[32, 81, 3] | fp32 | -2 | 28 | 0.0113 | 0.0107
[32, 16, 3] | fp32 | -2 | 28 | 0.0086 | 0.0078
[1, 81, 3] | fp32 | -2 | 28 | 0.0083 | 0.0078
[1, 16, 3] | fp32 | -2 | 28 | 0.0072 | 0.0071
[32, 81, 15130] | fp32 | -2 | 1 | 463.0909 | 115.262
[32, 16, 15130] | fp32 | -2 | 1 | 88.8961 | 12.469
[1, 81, 15130] | fp32 | -2 | 1 | 12.8996 | 1.7655
[1, 16, 15130] | fp32 | -2 | 1 | 2.5987 | 0.3342
[32, 81, 20] | fp32 | -2 | 1 | 0.5092 | 0.1445
[32, 16, 20] | fp32 | -2 | 1 | 0.1215 | 0.0437
[1, 81, 20] | fp32 | -2 | 1 | 0.0182 | 0.0069
[1, 16, 20] | fp32 | -2 | 1 | 0.0056 | 0.0034
[32, 81, 16] | fp32 | -2 | 1 | 0.4094 | 0.0541
[32, 16, 16] | fp32 | -2 | 1 | 0.0894 | 0.0142
[1, 81, 16] | fp32 | -2 | 1 | 0.0151 | 0.0039
[1, 16, 16] | fp32 | -2 | 1 | 0.0051 | 0.0027
[32, 81, 3] | fp32 | -2 | 1 | 0.0781 | 0.0693
[32, 16, 3] | fp32 | -2 | 1 | 0.018 | 0.0193
[1, 81, 3] | fp32 | -2 | 1 | 0.0047 | 0.0046
[1, 16, 3] | fp32 | -2 | 1 | 0.0027 | 0.0029
[32, 81, 15130] | fp64 | -2 | 28 | 40.4192 | 15.3224
[32, 16, 15130] | fp64 | -2 | 28 | 8.9532 | 3.1854
[1, 81, 15130] | fp64 | -2 | 28 | 3.1217 | 0.4471
[1, 16, 15130] | fp64 | -2 | 28 | 0.1816 | 0.0261
[32, 81, 20] | fp64 | -2 | 28 | 0.0395 | 0.0271
[32, 16, 20] | fp64 | -2 | 28 | 0.0139 | 0.0211
[1, 81, 20] | fp64 | -2 | 28 | 0.0317 | 0.0368
[1, 16, 20] | fp64 | -2 | 28 | 0.0126 | 0.0139
[32, 81, 16] | fp64 | -2 | 28 | 0.0341 | 0.032
[32, 16, 16] | fp64 | -2 | 28 | 0.0132 | 0.013
[1, 81, 16] | fp64 | -2 | 28 | 0.0255 | 0.0303
[1, 16, 16] | fp64 | -2 | 28 | 0.0114 | 0.0132
[32, 81, 3] | fp64 | -2 | 28 | 0.0126 | 0.0144
[32, 16, 3] | fp64 | -2 | 28 | 0.0083 | 0.0103
[1, 81, 3] | fp64 | -2 | 28 | 0.0088 | 0.0102
[1, 16, 3] | fp64 | -2 | 28 | 0.0068 | 0.0088
[32, 81, 15130] | fp64 | -2 | 1 | 724.4058 | 172.5648
[32, 16, 15130] | fp64 | -2 | 1 | 143.4831 | 26.8466
[1, 81, 15130] | fp64 | -2 | 1 | 18.7121 | 2.4747
[1, 16, 15130] | fp64 | -2 | 1 | 3.7613 | 0.4123
[32, 81, 20] | fp64 | -2 | 1 | 0.7511 | 0.2147
[32, 16, 20] | fp64 | -2 | 1 | 0.1659 | 0.0556
[1, 81, 20] | fp64 | -2 | 1 | 0.0271 | 0.0089
[1, 16, 20] | fp64 | -2 | 1 | 0.0077 | 0.0038
[32, 81, 16] | fp64 | -2 | 1 | 0.6073 | 0.0678
[32, 16, 16] | fp64 | -2 | 1 | 0.1232 | 0.0155
[1, 81, 16] | fp64 | -2 | 1 | 0.021 | 0.0043
[1, 16, 16] | fp64 | -2 | 1 | 0.0061 | 0.0027
[32, 81, 3] | fp64 | -2 | 1 | 0.1111 | 0.1089
[32, 16, 3] | fp64 | -2 | 1 | 0.025 | 0.0249
[1, 81, 3] | fp64 | -2 | 1 | 0.0057 | 0.0055
[1, 16, 3] | fp64 | -2 | 1 | 0.0028 | 0.0028

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - thanks, those numbers look great!

The [test failures](https://app.circleci.com/pipelines/github/pytorch/pytorch/329728/workflows/4e9abf12-3c0b-4e4b-95f7-08c413ef3b07/jobs/13833140/tests#failed-test-0) seem legitimate, in particular I'm seeing

```
test_torch_nn_Softmax2d
test_torch_nn_Softmin_multidim
test_torch_nn_functional_softmax_functional_dim0
test_torch_nn_functional_softmax_spatial
test_torch_nn_functional_softmax_spatial_dtype
```

fail with a few nan comparisons.

**Comment by leslie-fang-intel** [comment]
> @leslie-fang-intel - thanks, those numbers look great!
> 
> The [test failures](https://app.circleci.com/pipelines/github/pytorch/pytorch/329728/workflows/4e9abf12-3c0b-4e4b-95f7-08c413ef3b07/jobs/13833140/tests#failed-test-0) seem legitimate, in particular I'm seeing
> 
> ```
> test_torch_nn_Softmax2d
> test_torch_nn_Softmin_multidim
> test_torch_nn_functional_softmax_functional_dim0
> test_torch_nn_functional_softmax_spatial
> test_torch_nn_functional_softmax_spatial_dtype
> ```
> 
> fail with a few nan comparisons.

Thanks, I will look into these failures.

**Comment by leslie-fang-intel** [comment]
@cpuhrsch I think most of the failures has been resolved. The left 2 failures may relate with the code base. I will rebase and try again. And I have updated the performance data in above tables with dtype of double.

**Comment by leslie-fang-intel** [comment]
Emmm.... It seems `Facebook Internal - Builds & Tests` fails. But I am afraid I am unable to check the details @cpuhrsch

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - I reimported the diff on a newer base. The failures looked unrelated.

**Comment by leslie-fang-intel** [comment]
> @leslie-fang-intel - I reimported the diff on a newer base. The failures looked unrelated.
@cpuhrsch , It seems the internal test fails again 😂

**Review by cpuhrsch — APPROVED** [review]
Internal test is unrelated. Thank you for your work @leslie-fang-intel!
